monty-31/src/x86_64_avx2/packing.rs

```diff
@@ -227,6 +277,82 @@ fn mul<MPAVX2: MontyParametersAVX2>(lhs: __m256i, rhs: __m256i) -> __m256i {
+/// Inputs must be signed 32-bit integers.
+/// Outputs will be a signed integer in (-P, ..., P) copied into both the even and odd indices.
+#[inline]
+#[must_use]
+fn shifted_square<MPAVX2: MontyParametersAVX2>(input: __m256i) -> __m256i {
+    // Note that we do not need a restriction on the size of input[i]^2 as
+    // 2^30 < P and |i32| <= 2^31 and so => input[i]^2 <= 2^62 < 2^32P.
+    unsafe {
+        let square = x86_64::_mm256_mul_epi32(input, input);
+        let square_red = monty_red_unsigned::<MPAVX2>(square);
+        movehdup_epi32(square_red)
+    }
+}
+
+/// Cube the MontyField31 field elements in the even index entries.
+/// Inputs must be signed 32-bit integers in [-P, ..., P].
+/// Outputs will be a signed integer in (-P, ..., P) stored in the odd indices.
+#[inline]
+#[must_use]
+fn packed_exp_3<MPAVX2: MontyParametersAVX2>(input: __m256i) -> __m256i {
+    let square = shifted_square::<MPAVX2>(input);
+    monty_d_signed::<MPAVX2>(square, input)
+}
+
+/// Take the fifth power of the MontyField31 field elements in the even index entries.
+/// Inputs must be signed 32-bit integers in [-P, ..., P].
+/// Outputs will be a signed integer in (-P, ..., P) stored in the odd indices.
+#[inline]
+#[must_use]
+fn packed_exp_5<MPAVX2: MontyParametersAVX2>(input: __m256i) -> __m256i {
+    let square = shifted_square::<MPAVX2>(input);
+    let quad = shifted_square::<MPAVX2>(square);
+    monty_d_signed::<MPAVX2>(quad, input)
+}
+
+/// Take the seventh power of the MontyField31 field elements in the even index entries.
+/// Inputs must lie in [-P, ..., P].
+/// Outputs will also lie in (-P, ..., P) stored in the odd indices.
+#[inline]
+#[must_use]
+fn packed_exp_7<MPAVX2: MontyParametersAVX2>(input: __m256i) -> __m256i {
+    let square = shifted_square::<MPAVX2>(input);
+    let cube = monty_d_signed::<MPAVX2>(square, input);
+    let cube_shifted = movehdup_epi32(cube);
+    let quad = shifted_square::<MPAVX2>(square);
+
+    monty_d_signed::<MPAVX2>(quad, cube_shifted)
+}
+
+/// Apply func to the even and odd indices of the input vector.
+/// func should only depend in the 32 bit entries in the even indices.
+/// The output of func must lie in (-P, ..., P) and be stored in the odd indices.
+/// The even indices of the output of func will not be read.
+/// the input should should conform to the requirements of func
+#[inline]
+#[must_use]
+fn apply_func_to_even_odd<MPAVX2: MontyParametersAVX2>(
+    input: __m256i,
+    func: fn(__m256i) -> __m256i,
+) -> __m256i {
+    unsafe {
+        let input_evn = input;
+        let input_odd = movehdup_epi32(input);
+
+        let d_evn = func(input_evn);
+        let d_odd = func(input_odd);
+
+        let d_evn_hi = movehdup_epi32(d_evn);
+        let t = x86_64::_mm256_blend_epi32::<0b10101010>(d_evn_hi, d_odd);
+
+        let u = x86_64::_mm256_add_epi32(t, MPAVX2::PACKED_P);
+        x86_64::_mm256_min_epu32(t, u)
+    }
+}
```

Comment:
Nit: Arguably simpler, but mainly just a stylistic thing: I'd probably make this whole function `unsafe` and put it in the (inevitable) unsafe block at the call sites.
```suggestion
unsafe fn apply_func_to_even_odd<MPAVX2: MontyParametersAVX2>(
    input: __m256i,
    func: fn(__m256i) -> __m256i,
) -> __m256i {
    let input_evn = input;
    let input_odd = movehdup_epi32(input);

    let d_evn = func(input_evn);
    let d_odd = func(input_odd);

    let d_evn_hi = movehdup_epi32(d_evn);
    let t = x86_64::_mm256_blend_epi32::<0b10101010>(d_evn_hi, d_odd);

    let u = x86_64::_mm256_add_epi32(t, MPAVX2::PACKED_P);
    x86_64::_mm256_min_epu32(t, u)
}
```